flight/PiOS/Common/pios_bmp280.c

```diff
@@ -0,0 +1,441 @@
+{
+
+	if (PIOS_BMP280_Validate(dev) != 0)
+		return -1;
+
+	const struct pios_i2c_txn txn_list[] = {
+		{
+			.info = __func__,
+			.addr = BMP280_I2C_ADDR,
+			.rw = PIOS_I2C_TXN_WRITE,
+			.len = 1,
+			.buf = &address,
+		}
+		,
+		{
+		 .info = __func__,
+		 .addr = BMP280_I2C_ADDR,
+		 .rw = PIOS_I2C_TXN_READ,
+		 .len = len,
+		 .buf = buffer,
+		 }
+	};
+
+	return PIOS_I2C_Transfer(dev->i2c_id, txn_list, NELEMENTS(txn_list));
+}
+
+/**
+* Writes one or more bytes to the BMP280
+* \param[in] address Register address
+* \param[in] buffer source buffer
+* \return 0 if operation was successful
+* \return -1 if error during I2C transfer
+*/
+static int32_t PIOS_BMP280_WriteCommand(uint8_t address, uint8_t buffer)
+{
+	uint8_t data[] = {
+		address,
+		buffer,
+	};
+
+	if (PIOS_BMP280_Validate(dev) != 0)
+		return -1;
+
+	const struct pios_i2c_txn txn_list[] = {
+		{
+		 .info = __func__,
+		 .addr = BMP280_I2C_ADDR,
+		 .rw = PIOS_I2C_TXN_WRITE,
+		 .len = sizeof(data),
+		 .buf = data,
+		 }
+		,
+	};
+
+	return PIOS_I2C_Transfer(dev->i2c_id, txn_list, NELEMENTS(txn_list));
+}
+
+/**
+* @brief Run self-test operation.
+* \return 0 if self-test succeed, -1 if failed
+*/
+int32_t PIOS_BMP280_Test()
+{
+	if (PIOS_BMP280_Validate(dev) != 0)
+		return -1;
+
+	// TODO: Is there a better way to test this than just checking that pressure/temperature has changed?
+	int32_t currentTemperature = 0;
+	int32_t currentPressure    = 0;
+
+	PIOS_BMP280_ClaimDevice();
+
+	currentTemperature = dev->compensatedTemperature;
+	currentPressure    = dev->compensatedPressure;
+
+	PIOS_BMP280_StartADC();
+	PIOS_DELAY_WaitmS(PIOS_BMP280_GetDelay());
+	PIOS_BMP280_ReadADC(true);
+	PIOS_BMP280_ReleaseDevice();
+
+	if (currentTemperature == dev->compensatedTemperature)
+		return -1;
+
+	if (currentPressure == dev->compensatedPressure)
+		return -1;
+
+	return 0;
+}
+
+static void PIOS_BMP280_Task(void *parameters)
+{
+	int32_t temp_press_interleave_count = dev->temperature_interleaving;
+	int32_t read_adc_result = 0;
+	bool    computeTemp = false;
+
+	while (1) {
+
+		temp_press_interleave_count --;
+
+		if(temp_press_interleave_count <= 0) {
+			computeTemp = true;
+			temp_press_interleave_count = dev->temperature_interleaving;
+		}
+		else
+			computeTemp = false;
+
+		PIOS_BMP280_ClaimDevice();
+		PIOS_BMP280_StartADC();
+		PIOS_Thread_Sleep(PIOS_BMP280_GetDelay());
+		read_adc_result = PIOS_BMP280_ReadADC(computeTemp);
+		PIOS_BMP280_ReleaseDevice();
+
+		// Compute the altitude from the pressure and temperature and send it out
+		struct pios_sensor_baro_data data;
+		data.temperature = ((float) dev->compensatedTemperature) / 100.0f;
+		data.pressure = ((float) dev->compensatedPressure) / 256.0f / 1000.0f;
+		data.altitude = 44330.0f * (1.0f - powf(data.pressure / BMP280_P0, (1.0f / 5.255f)));
+
+		if (read_adc_result == 0) {
+			PIOS_Queue_Send(dev->queue, (void*)&data, 0);
+		}
+	}
+}
+
+#endif /* PIOS_INCLUDE_BMP280 */
```

Comment:
Don't know it will work with others.  I heard rumors the BMP180 and BMP280 are similar, but would need to research it.